opengovsg/postmangovsg
# chore: replace local module with shared folder

## Problem

Closes #1139 

## Solution

### Use Typescript project references to reference shared module
I've chosen to take a slightly different approach to referencing the shared module in `tsconfig` as compared to the [vaccinegovsg-clinic](https://github.com/datagovsg/vaccinegovsg-clinic/tree/dev/frontend) repository. Instead of adding the shared folder path in the `include` option, I've chosen to make use of [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) instead. 

The main advantage for this approach is that we can avoid having to manually keep in sync the dependencies for the `shared` folder and the other modules.  For example, if we were to use the `include` approach, the `xss` library has to be in each of the `package.json` for `backend`, `frontend` and `worker`. Keeping all the different `package.json` in sync is a non-trivial problem and is prone to developer error.

The drawback of this approach is that we have to manually `npm install` inside the shared folder to install its dependencies. The necessary changes to `.travis.yml` and the respective `Dockerfile` have been made.

### Manually patch `moduleNameMapper` rather than use `aliasJest`
The default `aliasJest` function provided by [react-app-rewire-alias](https://github.com/oklas/react-app-rewire-alias) does not work for our setup. The main reason for this is that the library assumes that our aliases would follow the pattern of `module/submodule`. However, there are quite a number of instances in our code that references aliases directly without a submodule (e.g. `import config from 'config'`).

## Tests
Apart from changing the structure of the code, this PR should not modify or introduce any behaviour. To verify that everything works as expected:
1. Ensure all tests for backend and frontend passes
2. Ensure deployment succeeds without errors for backend and frontend

In addition, hot reload should also work now for the shared module. Changing code in the shared module should automatically trigger a recompilation for any other module depending on it.
